test-data/unit/check-protocols.test

```diff
@@ -3546,3 +3546,142 @@ S = TypeVar("S")
 b: Type[B]
 reveal_type(test(b))  # N: Revealed type is "__main__.B"
+
+[case testModuleAsProtocolImplementation]
+import default_config
+import bad_config_1
+import bad_config_2
+import bad_config_3
+from typing import Protocol
+
+class Options(Protocol):
+    timeout: int
+    one_flag: bool
+    other_flag: bool
+    def update(self) -> bool: ...
+
+def setup(options: Options) -> None: ...
+setup(default_config)  # OK
+setup(bad_config_1)  # E: Argument 1 to "setup" has incompatible type Module; expected "Options" \
+                     # N: "ModuleType" is missing following "Options" protocol member: \
+                     # N:     timeout
+setup(bad_config_2)  # E: Argument 1 to "setup" has incompatible type Module; expected "Options" \
+                     # N: Following member(s) of "Module bad_config_2" have conflicts: \
+                     # N:     one_flag: expected "bool", got "int"
+setup(bad_config_3)  # E: Argument 1 to "setup" has incompatible type Module; expected "Options" \
+                     # N: Following member(s) of "Module bad_config_3" have conflicts: \
+                     # N:     Expected: \
+                     # N:         def update() -> bool \
+                     # N:     Got: \
+                     # N:         def update(obj: Any) -> bool
+
+[file default_config.py]
+timeout = 100
+one_flag = True
+other_flag = False
+def update() -> bool: ...
+
+[file bad_config_1.py]
+one_flag = True
+other_flag = False
+def update() -> bool: ...
+
+[file bad_config_2.py]
+timeout = 100
+one_flag = 42
+other_flag = False
+def update() -> bool: ...
+
+[file bad_config_3.py]
+timeout = 100
+one_flag = True
+other_flag = False
+def update(obj) -> bool: ...
+[builtins fixtures/module.pyi]
+
+[case testModuleAsProtocolImplementationInference]
+import default_config
+from typing import Protocol, TypeVar
+
+T = TypeVar("T", covariant=True)
+class Options(Protocol[T]):
+    timeout: int
+    one_flag: bool
+    other_flag: bool
+    def update(self) -> T: ...
+
+def setup(options: Options[T]) -> T: ...
+reveal_type(setup(default_config))  # N: Revealed type is "builtins.str"
+
+[file default_config.py]
+timeout = 100
+one_flag = True
+other_flag = False
+def update() -> str: ...
+[builtins fixtures/module.pyi]
+
+[case testModuleAsProtocolImplementationClassObject]
+import runner
+import bad_runner
+from typing import Callable, Protocol
+
+class Runner(Protocol):
+    @property
+    def Run(self) -> Callable[[int], Result]: ...
+
+class Result(Protocol):
+    value: int
+
+def run(x: Runner) -> None: ...
+run(runner)  # OK
+run(bad_runner)  # E: Argument 1 to "run" has incompatible type Module; expected "Runner" \
+                 # N: Following member(s) of "Module bad_runner" have conflicts: \
+                 # N:     Expected: \
+                 # N:         def (int, /) -> Result \
+                 # N:     Got: \
+                 # N:         def __init__(arg: str) -> Run
+
+[file runner.py]
```

Comment:
I think it makes sence to test something like:

```python
from typing_extensions import Protocol
class My(Protocol):
   a: int  # is settable attr

def func(arg: My): pass

import some_module
func(some_module)  # error
```

And:

```python
# some_module.py
from typing_extensions import Final

a: Final = 1  # is module-level constant
```

Looks like we should correctly raise an error here.